# Chapter 04 조건문과 반복문 

## 01. 조건문 

## 02. Lab : 어떤 종류의 학생인지 맞히기 

## 03. 반복문 

## 04. Lab : 구구단 계산기 

## 05. 조건문과 반복문 실습 

## 06. Lab : 숫자 찾기 게임 

## 07. Lab : 연속적인 구구단 계산기

## 08. Lab : 평균 구하기 

## 09. 코드의 오류를 처리하는 방법 

## 요약 

## 연습문제 

**1. 주어진 자연수 N에 대해 N이 짝수이면 N!을, 홀수이면 ∑N을 구하는 코드를 작성하려 한다. 주어진 코드에서 잘못된 부분을 수정하시오.**

In [6]:
def calculator(N):
    if N % 2 == 0:
        result = 1
        for i in range(1, N+1):
            result = result * i
    else:
        result = 0
        for i in range(1, N+1):
            result = result + i
    return result

if __name__ == '__main__':
    print(calculator(3))

6


우선 calculator 함수의 인자는 N이고, 조건문에서 사용하는 변수는 n임.\
이 둘을 둘다 대문자로 하던지, 소문자로 하던지 맞춰주어야 문법적인 오류가 발생하지 않음.\
또한, range문은 (시작번호, 마지막번호)에서 마지막번호-1까지만 계산이 이루어지므로, N!이나 ∑N을 구하려면, 범위를 N+1까지 해주어야 논리적 오류가 발생하지 않는다

**2. 다음 코드의 실행 결과를 쓰시오.**

In [8]:
fruit = 'apple'
if fruit == 'Apple':
    fruit = 'Apple'
elif fruit == 'fruit':
    fruit = 'fruit'
else:
    fruit = fruit
print(fruit)

apple


fruit의 값은 처음에 'apple'로 초기화 되었다\
if문에 해당하는 'Apple'에도 해당하지 않고, elif문에 해당하는 'fruit'에도 해당하지 않으므로, else문으로 넘어가, 그대로 fruit의 값이 유지된다\
그러므로 fruit을 출력하면 apple값이 나온다

**3. 다음 코드의 실행 결과를 쓰시오.**

In [9]:
num = ['12', '34', '56']
for i in num:
    i = int(i)
print(num)

['12', '34', '56']


결과적으로 num값을 출력하면 되는 문제\
반복문에서 num값을 변화시키지 않음\
그러므로, 그대로 num값 ['12', '34','56']을 출력하면 됨 

**4. 다음 코드의 실행 결과를 쓰시오.**

In [10]:
number = ["1", 2, 3, float(4), str(5)]
if number[4] == 5:
    print(type(number[0]))
elif number[3] == 4:
    print(number[2:-1])

[3, 4.0]


number[4] == '5'\
'5' != 5 (문자형과 숫자형은 비교할 수 없다!)
number[3] == 4\
4(float형) == 4\
number[2:-1] = [3, float(4)] = [3, 4.0]

**5. 다음 코드의 실행 결과를 쓰시오.**

In [11]:
num = 0
i = 1

while i < 8:
    if i % 3 == 0:
        break
    i += 1
    num += i
print(num)

5


num은 0으로 i는 1로 초기화\
i = 1부터 while문 시작\
1 % 3 = 1\
i = 2\
num = 2\
i = 2로 반복됨\
2 % 3 = 2\
i = 2 + 1 = 3\
num = 2 + 3 = 5

**6. 다음 코드의 실행 결과를 쓰시오.**

In [12]:
result = 0
for i in range(5, -5, -2):
    if i < -3:
        result += 1
    else :
        result -= 1

print(result)

-5


for i in range(5, -5, -2)\
5, 3, 1, -1, -3 순으로 반복\
다 -3보다는 작지 않으므로, if문 조건에는 해당 안 됨\
else문만 5번 계산되므로 result는 -5

**7. 다음 코드의 실행 결과를 쓰시오.**\
2번 문제와 동일!

**8. 다음 코드의 실행 결과를 쓰시오.**

In [13]:
first_value = 0
second_value = 0
for i in range(1, 10):
    if i is 5:
        continue
        first_value = i
    if i is 10:
        break
        second_value = i

print(first_value + second_value)

0


<>:4: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:7: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:4: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:7: SyntaxWarning: "is" with a literal. Did you mean "=="?
C:\Users\User\AppData\Local\Temp/ipykernel_4664/2402932925.py:4: SyntaxWarning: "is" with a literal. Did you mean "=="?
  if i is 5:
C:\Users\User\AppData\Local\Temp/ipykernel_4664/2402932925.py:7: SyntaxWarning: "is" with a literal. Did you mean "=="?
  if i is 10:


for i in range(1, 10)으로, 1부터 9까지 i가 반복된다\
5일때와 10일때 조건문의 조건을 각각 충족하게 되는데,\
5의 경우 continue문을 받고 first_value = i를 만나지 못한채로 i는 6으로 넘어가게 된다\
그리고 if i is 10 구문은 i가 9까지만 반복되므로 만족하는 조건을 만나지 못한다\
그러므로 first_value와 second_value는 둘 다 0의 값을 갖는다

**9. 다음 코드의 실행 결과를 쓰시오.**

In [14]:
num = ""
for i in range(10):
    if i <= 5 and (i % 2) == 0:
        continue
    elif i is 7 or i is 10:
        continue
    else:
        num = str(i) + num
print(num)

986531


<>:5: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:5: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:5: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:5: SyntaxWarning: "is" with a literal. Did you mean "=="?
C:\Users\User\AppData\Local\Temp/ipykernel_4664/2311581283.py:5: SyntaxWarning: "is" with a literal. Did you mean "=="?
  elif i is 7 or i is 10:
C:\Users\User\AppData\Local\Temp/ipykernel_4664/2311581283.py:5: SyntaxWarning: "is" with a literal. Did you mean "=="?
  elif i is 7 or i is 10:


반복은 i가 0부터 9까지\
if문은 i가 5이하인 짝수거나 0의 경우 충족\
elif문은 i가 7이거나 10인 경우 충족\
위 두 조건을 만족하지 않으면 else문으로 내려와서, str형으로 바뀐뒤 문자열 앞부분으로 추가됨\
따라서 986531

**10. 다음 함수는 작업의 상태를 나타내는 함수이다. 코드의 실행 결과를 쓰시오.**

In [16]:
def work_status(task, worker, day):
    rest_task = task
    for k in range(day):
        if rest_task > 0:
            rest_task = rest_task - worker
        elif rest_task <= 0:
            print("Task end")
    if rest_task > 0:
        print("Hire more workers")

work_status(100, 11, 10)

In [17]:
work_status(100, 1, 10)

Hire more workers


In [18]:
work_status(100, 9, 10)

Hire more workers


In [19]:
work_status(100, 10, 10)

rest_task = 100\
k = 0 ~ 9
1) (100, 11, 10)\
k = 0, rest_task = 89\
k = 1, rest_task = 78\
k = 2, rest_task = 67\
k = 3, rest_task = 56\
k = 4, rest_task = 45\
k = 5, rest_task = 34\
k = 6, rest_task = 23\
k = 7, rest_task = 12\
k = 8, rest_task = 1\
k = 9, rest_task = -10

2) (100, 1, 10)
k = 0, rest_task = 99\
k = 1, rest_task = 98\
k = 2, rest_task = 97\
k = 3, rest_task = 96\
k = 4, rest_task = 95\
k = 5, rest_task = 94\
k = 6, rest_task = 93\
k = 7, rest_task = 92\
k = 8, rest_task = 91\
k = 9, rest_task = 90\
rest_task > 0 -> Hire more workers

3) (100, 9, 10)
k = 0, rest_task = 91\
k = 1, rest_task = 82\
k = 2, rest_task = 73\
k = 3, rest_task = 64\
k = 4, rest_task = 55\
k = 5, rest_task = 46\
k = 6, rest_task = 37\
k = 7, rest_task = 28\
k = 8, rest_task = 19\
k = 9, rest_task = 10\
rest_task > 0 -> Hire more workers

4) (100, 10, 10)
k = 0, rest_task = 90\
k = 1, rest_task = 80\
k = 2, rest_task = 70\
k = 3, rest_task = 60\
k = 4, rest_task = 50\
k = 5, rest_task = 40\
k = 6, rest_task = 30\
k = 7, rest_task = 20\
k = 8, rest_task = 10\
k = 9, rest_task = 0

**11. 다음 코드의 실행 결과를 쓰시오.**

In [33]:
score_list = [5, 10, 15, 20, 25, 30]

sum_of_score = 0
i = 0
while i < len(score_list):
    if i % 2 == 0:
        sum_of_score += score_list[i]
    i += 1
    
print(sum_of_score)

45


len(score_list) = 6\
i = 0 ~ 5\
i가 0이거나 짝수일때 score_list의 값을 sum_of_score에 더해줌\
i = 0, 2, 4\
sum_of_score = 5 + 15 + 25 = 45

**12. 다음 코드의 실행 결과를 쓰시오.**

In [32]:
coupon = 0
money = 200000
coffee = 3500
while money > coffee:
    if coupon < 4:
        money = money - coffee
        coupon += 1
    else:
        money += 2800
        coupon = 0
print(money)

2600


coupon = 0, money = 200000, coffee = 3500\
200000 > 3500, coupon = 0\
money = 196500 coupon = 1\
money = 193000 coupon = 2\
money = 189500 coupon = 3\
money = 186000 coupon = 4\
money = 188800 coupon = 0\
쿠폰을 4개 모아 사용할때까지 총 11200(14000 - 2800)원 사용\
200000 / 11200 = 17 ... 9600\
money = 9600 coupon = 0\
money = 6100 coupon = 1\
money = 2600 coupon = 2\
남은 money는 2600원

**13. 다음 코드는 변수 a와 b의 숫자를 비교하여 숫자와 자릿수가 같으면 'strike', 숫자는 서로 포함하나 자릿수가 다르면 'ball'로 카운트하는 프로그램이다. 다음 프로그램의 실행 결과를 쓰시오. 또한, 결과값이 맞지 않는 이유를 쓰고, 올바른 결과값이 나오도록 코드를 수정하시오.**

In [30]:
a = "369"
b = "693"

strike = 0
ball = 0

for number in a:
    if b.count(number) >= 1: # > 1
        if b.find(number) == a.find(number):
            strike  += 1 # != 
        else:
            ball += 1

print("Strike:", strike, "Ball:", ball)

Strike: 0 Ball: 3


a와 b를 비교했을 때, 숫자와 자릿수가 같으면 strike, 같은 숫자는 포함하나 그 자릿수가 다르면 ball로 카운트\
a와 b는 각각 문자열형으로 369, 693을 저장\
strike와 ball에는 0의 값으로 초기화\
문자열 a의 문자요소 하나씩을 반복. 3 -> 6 -> 9\
가장 먼저 if문에 count()함수를 사용하여 b에 a의 해당 요소가 있는지를 확인\
해당 요소가 있는 경우는 1 이상이기만 하면 되므로 > 1이 아니라 >= 1로 바꾸어줘야함\
1이상의 값이 반환된 경우(b에 a의 해당 요소가 있는 경우), find()함수를 이용해서 둘의 자릿수가 같은지 확인\
여기서 숫자와 자릿수가 같다면 strike이므로 !=가 아니라 == 이어야함!

**14. 다음과 같이 코드를 작성했을 때, 실행 결과로 알맞은 것은?**

In [21]:
list_data_a = [1, 2]
list_data_b = [3, 4]

for i in list_data_a:
    for j in list_data_b:
        result = i + j
print(result)

6


i는 list_data_a의 요소인 1, 2가 올 것\
j는 list_data_b의 요소인 3, 4가 올 것\
result에는 먼저 1과 3이 더해진 4, 1과 4가 더해진 5,\
2와 3이 더해진 5, 2와 4가 더해진 6이 올 것\
그러나 result는 리스트가 아니라 그냥 하나의 값을 갖는 변수라 마지막 6만 result값이 되어 반복문이 종료됨\
그러므로 답은 6

**15. 다음과 같이 코드를 작성했을 때, 실행 결과로 알맞은 것은?**

1번. [1, 2, 3, 4, 5, 6]

In [20]:
list_1 = [[1, 2], [3], [4, 5, 6]]
a, b, c = list_1
list_2 = a + b + c

print(list_2)

[1, 2, 3, 4, 5, 6]


중첩 리스트 list_1을 a, b, c로 언패킹했다가 다시 리스트 덧셈연산을 이용하여 list_2라는 하나의 새로운 리스트로 만들어 줌.